qutebrowser/misc/keyhintwidget.py

```diff
@@ -0,0 +1,115 @@
+
+When a user inputs a key that forms a partial match, this shows a small window
+with each possible completion of that keystring and the corresponding command.
+It is intended to help discoverability of keybindings.
+"""
+
+import html
+
+from PyQt5.QtWidgets import QLabel, QSizePolicy
+from PyQt5.QtCore import pyqtSlot, pyqtSignal
+
+from qutebrowser.config import config, style
+from qutebrowser.utils import objreg, utils
+
+
+class KeyHintView(QLabel):
+
+    """The view showing hints for key bindings based on the current key string.
+
+    Attributes:
+        _win_id: Window ID of parent.
+        _enabled: If False, do not show the window at all
+        _suffix_color: Highlight for completions to the current keychain.
+
+    Signals:
+        reposition_keyhint: Emitted when this widget should be resized.
+    """
+
+    STYLESHEET = """
+        QLabel {
+            font: {{ font['keyhint'] }};
+            color: {{ color['keyhint.fg'] }};
+            background-color: {{ color['keyhint.bg'] }};
+            padding: 6px;
+            padding-bottom: -4px;
+            border-radius: 6px;
+        }
+    """
+
+    reposition_keyhint = pyqtSignal()
+
+    def __init__(self, win_id, parent=None):
+        super().__init__(parent)
+        self._win_id = win_id
+        self.set_enabled()
+        cfg = objreg.get('config')
+        cfg.changed.connect(self.set_enabled)
+        style.set_register_stylesheet(self)
+        self.setSizePolicy(QSizePolicy.Fixed, QSizePolicy.Minimum)
+        self.hide()
+
+    def __repr__(self):
+        return utils.get_repr(self, win_id=self._win_id)
+
+    @config.change_filter('ui', 'show-keyhints')
+    def set_enabled(self):
+        """Update self._enabled when the config changed."""
+        self._enabled = config.get('ui', 'show-keyhints')
+        if not self._enabled:
+            self.hide()
+
+    def showEvent(self, e):
+        """Adjust the keyhint size when it's freshly shown."""
+        self.reposition_keyhint.emit()
+        super().showEvent(e)
+
+    @pyqtSlot(str)
+    def update_keyhint(self, modename, prefix):
+        """Show hints for the given prefix (or hide if prefix is empty).
+
+        Args:
+            prefix: The current partial keystring.
+        """
+        if not prefix or not self._enabled:
+            self.hide()
+            return
+
+        self.show()
+        suffix_color = html.escape(config.get('colors', 'keyhint.fg.suffix'))
+
+        text = ''
+        keyconf = objreg.get('key-config')
+        # this is only fired in normal mode
+        for key, cmd in keyconf.get_bindings_for(modename).items():
+            if key.startswith(prefix):
+                suffix = "<font color='{}'>{}</font>".format(suffix_color,
+                    html.escape(key[len(prefix):]))
+
+                text += '{}{}\t<b>{}</b><br>'.format(html.escape(prefix),
+                                                     suffix,
```

Comment:
The args used to format suffix are already escaped. `x<` would be escaped in the
`suffix =` statement. If I escape `suffix` itself, _then_ it will be invalid
html as the legitimate html tags will be escaped.